torchvision/datasets/video_utils.py

```diff
@@ -0,0 +1,190 @@
+                 frame_rate=None):
+        self.video_paths = video_paths
+        self._compute_frame_pts()
+        self.compute_clips(clip_length_in_frames, frames_between_clips, frame_rate)
+
+    def _compute_frame_pts(self):
+        self.video_pts = []
+        self.video_fps = []
+        # TODO maybe paralellize this
+        for video_file in self.video_paths:
+            clips, fps = read_video_timestamps(video_file)
+            self.video_pts.append(torch.as_tensor(clips))
+            self.video_fps.append(fps)
+
+    def compute_clips(self, num_frames, step, frame_rate=None):
+        """
+        Compute all consecutive sequences of clips from video_pts.
+        Always returns clips of size `num_frames`, meaning that the
+        last few frames in a video can potentially be dropped.
+
+        Arguments:
+            num_frames (int): number of frames for the clip
+            step (int): distance between two clips
+            dilation (int): distance between two consecutive frames
+                in a clip
+        """
+        self.num_frames = num_frames
+        self.step = step
+        self.frame_rate = frame_rate
+        self.clips = []
+        for video_pts, fps in zip(self.video_pts, self.video_fps):
+            if frame_rate is not None:
+                # divup, == int(ceil(fps / frame_rate))
+                dilation = max(int((fps + frame_rate - 1) // frame_rate), 1)
+            else:
+                dilation = 1
+            clips = unfold(video_pts, num_frames, step, dilation)
+            self.clips.append(clips)
+        clip_lengths = torch.as_tensor([len(v) for v in self.clips])
+        self.cumulative_sizes = clip_lengths.cumsum(0).tolist()
+
+    def __len__(self):
+        return self.num_clips()
+
+    def num_videos(self):
+        return len(self.video_paths)
+
+    def num_clips(self):
+        """
+        Number of subclips that are available in the video list.
+        """
+        return self.cumulative_sizes[-1]
+
+    def get_clip_location(self, idx):
+        """
+        Converts a flattened representation of the indices into a video_idx, clip_idx
+        representation.
+        """
+        video_idx = bisect.bisect_right(self.cumulative_sizes, idx)
+        if video_idx == 0:
+            clip_idx = idx
+        else:
+            clip_idx = idx - self.cumulative_sizes[video_idx - 1]
+        return video_idx, clip_idx
+
+    def resample_video(self, video, original_fps, new_fps):
+        """
+        Resamples a video so that it matches a new fps specified
+        """
+        step = original_fps / new_fps
+        if step.is_integer():
+            # optimization: if step is integer, don't need to perform
+            # advanced indexing
+            step = int(step)
+            return video[::step]
+        idxs = torch.arange(self.num_frames, dtype=torch.float32) * step
+        idxs = idxs.floor().to(torch.int64)
```

Comment:
I think it doesn't match the way `dilation` is computed on line 83.

utest
```
video_pts = torch.arange(30)
fps = 30
frame_rate = 13.0
size = 10
dilation = max(int((fps + frame_rate - 1) // frame_rate), 1)  # = 3
clips = unfold(video_pts, size, size, dilation)  # = tensor([[ 0,  3,  6,  9, 12, 15, 18, 21, 24, 27]])

step = fps / frame_rate
idxs = torch.arange(size, dtype=torch.float32) * step
idxs = idxs.floor().to(torch.int64)
video_pts[idxs]  # =tensor([ 0,  2,  4,  6,  9, 11, 13, 16, 18, 20]), whereas it should match clips[0]
```
I think the current code works only if `fps / frame_rate` is (almost) an integer

Reply:
This is a great point, once again, thanks Lowik!

So, here is my view of things:

you have 1 second of video at 30 fps. We now resample it to be at 13fps. Now we have 13 frames.
Their time coordinates are given by
```python
torch.arange(13, dtype=torch.float32) * 30 / 13
# tensor([ 0.0000,  2.3077,  4.6154,  6.9231,  9.2308, 11.5385, 13.8462, 16.1538,
#        18.4615, 20.7692, 23.0769, 25.3846, 27.6923])
```
Now, we want to take 10 frames out of this newly sampled video. This gives us
```
tensor([ 0.0000,  2.3077,  4.6154,  6.9231,  9.2308, 11.5385, 13.8462, 16.1538,
        18.4615, 20.7692])
```
Because we only do nearest interpolation on the frames, this corresponds to (by calling `.floor()`
```
tensor([ 0,  2,  4,  6,  9, 11, 13, 16, 18, 20])
```
Note that the indices are actually exactly the same as what we originally compute
```python
idxs = torch.arange(10, dtype=torch.float32) * (30 / 13)
# tensor([ 0.0000,  2.3077,  4.6154,  6.9231,  9.2308, 11.5385, 13.8462, 16.1538,
#         18.4615, 20.7692])
```

So I'm inclined to say that the current behavior that we have, which returns the interpolation on the frames, is actually the behavior we want: interpolate to fps 13, then return 10 consecutive frames.

Thoughts?

Reply:
> So I'm inclined to say that the current behavior that we have, which returns the interpolation on the frames, is actually the behavior we want: interpolate to fps 13, then return 10 consecutive frames.

I  agree, although I wonder if the inconsistent gaps would allow networks to cheat in some way. I've never tried out the interpolation where the resampling was not done by integer factor, so I suppose we'll have to try it out.

Reply:
My current take on this is that we should do the most correct behavior as possible, and rescale the whole video first, and then perform the `unfold`.

This will be slightly less efficient, but much more predictable.